src/main/java/frc/robot/auto/PathFollower.java

```diff
@@ -0,0 +1,128 @@
+
+package frc.robot.auto;
+
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.geometry.Translation2d;
+import frc.robot.RobotContainer;
+import frc.robot.libraries.geometry.Line2d;
+
+public class PathFollower {
+
+    final Path path;
+    
+    int lastCrossedPointIndex = 0;
+
+    double lookAheadMeters; // Calculate-able?
+
+    /**
+     * 
+     * @param FollowedPath
+     */
+    public PathFollower(Path FollowedPath) {
+        path = FollowedPath;
+    }
+
+    public void setLookAheadDistance(double distanceMeters) {
+        lookAheadMeters = distanceMeters;
+    }
+
+    /**
+     * Calculates a PathState indicating where the robot should
+     * drive to, and how fast it should be traveling to that point.
+     * @param robotPosition current position of robot in meters
+     * @return State for robot to travel to from grabbed position along path.
+     */
+    public PathState getPathState(Pose2d robotPosition) {
+        PathPoint lastPoint = path.points.get(lastCrossedPointIndex);
+        PathPoint nextPoint; // Avoid init, though declare access
+        // Catch index out of bounds, thus we know its the end of the path.
+        try {
+            nextPoint = path.points.get(lastCrossedPointIndex + 1);
+        } catch (IndexOutOfBoundsException e) {
+            e.printStackTrace();
+            // lastCrossedPointIndex should never reach
+            // this, if it does, decrement index and recurse            
+            lastCrossedPointIndex --;
+            return getPathState(robotPosition);
+        }
+        
+        Translation2d lastPointPos = lastPoint.getPosMeters(), nextPointPos = nextPoint.getPosMeters();
+        // Line class to allow for the math to be outsourced, just making this code cleaner.
+        Line2d pathLine = new Line2d(lastPointPos, nextPointPos); // Line in original path coords
+        Line2d lineFromOrigin = pathLine.adjustToOrigin(); // Line adjusted to origin
+
+    
+        // Calculates the coordinates of the closest point upon the path to the robot
+        Translation2d robotPathIntersection = lineFromOrigin.findPerpendicularIntersection(
+            // Adjust robot position according to line adjustment
+            robotPosition.getTranslation().minus(lastPointPos)
+        );
+
+        // Construct distance and percent variables for both position, and lookahead
+        double robotDistanceAlongLineMeters = robotPathIntersection.getNorm();
+        double lookAheadDistanceAlongLineMeters = robotDistanceAlongLineMeters + lookAheadMeters;
+
+        // Lines of a length of 0 are invalid
+        // Normalize meters values
+        double percentFromLastPoint = robotDistanceAlongLineMeters / lineFromOrigin.getLength();
+        double lookAheadPercentFromLastPoint = lookAheadDistanceAlongLineMeters / lineFromOrigin.getLength();
+
+        // Check if lookahead is past this line
+        if (
```

Comment:
incrementing the last point based on on the look ahead makes the robot chase the last point crossed. The last point crossed should always be behind the robot.